packages/client/src/workflow-client.ts

```diff
@@ -834,26 +962,121 @@ export class WorkflowClient extends BaseClient {
       this.rethrowUpdateGrpcError(err, 'Workflow Update failed', input.workflowExecution);
     }
     return {
-      updateId,
+      updateId: request.request!.meta!.updateId!,
       // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
       workflowRunId: response.updateRef!.workflowExecution!.runId!,
       outcome: response.outcome ?? undefined,
     };
   }
 
+  /**
+   * Send the Update-With-Start MultiOperation request.
+   *
+   * Used as the final function of the interceptor chain during
+   * startUpdateWithStart and executeUpdateWithStart.
+   */
+  protected async _updateWithStartHandler(
+    waitForStage: WorkflowUpdateStage,
+    onStart: (startResponse: temporal.api.workflowservice.v1.IStartWorkflowExecutionResponse) => void,
+    input: WorkflowStartUpdateWithStartInput
+  ): Promise<WorkflowStartUpdateWithStartOutput> {
+    const startInput: WorkflowStartInput = {
+      workflowType: input.workflowType,
+      options: input.workflowStartOptions,
+      headers: input.workflowStartHeaders,
+    };
+    const updateInput: WorkflowStartUpdateInput = {
+      updateName: input.updateName,
+      args: input.updateArgs,
+      workflowExecution: {
+        workflowId: input.workflowStartOptions.workflowId,
+      },
+      options: input.updateOptions,
+      headers: input.updateHeaders,
+    };
+    const startRequest = await this.createStartWorkflowRequest(startInput);
+    const waitForStageProto = encodeWorkflowUpdateStage(waitForStage)!;
+    const updateRequest = await this._createUpdateWorkflowRequest(waitForStageProto, updateInput);
+    const multiOpReq: temporal.api.workflowservice.v1.IExecuteMultiOperationRequest = {
+      namespace: this.options.namespace,
+      operations: [
+        {
+          startWorkflow: startRequest,
+        },
+        {
+          updateWorkflow: updateRequest,
+        },
+      ],
+    };
+
+    let multiOpResp: temporal.api.workflowservice.v1.IExecuteMultiOperationResponse;
+    let startResp: temporal.api.workflowservice.v1.IStartWorkflowExecutionResponse;
+    let updateResp: temporal.api.workflowservice.v1.IUpdateWorkflowExecutionResponse;
+    let reachedStage: temporal.api.enums.v1.UpdateWorkflowExecutionLifecycleStage;
+    do {
+      try {
+        multiOpResp = await this.workflowService.executeMultiOperation(multiOpReq);
+        // TODO: order is guaranteed but check structural validity of response,
+        // e.g. startWorkflow / updateWorkflow keys
+        startResp = multiOpResp.responses?.[0]
+          ?.startWorkflow as temporal.api.workflowservice.v1.IStartWorkflowExecutionResponse;
+        onStart(startResp);
+        updateResp = multiOpResp.responses?.[1]
+          ?.updateWorkflow as temporal.api.workflowservice.v1.IUpdateWorkflowExecutionResponse;
+        reachedStage =
+          updateResp.stage ??
+          UpdateWorkflowExecutionLifecycleStage.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED;
+      } catch (err) {
```

Comment:
I think moving the while-loop (and probably the var definitions and the return statement) _inside_ of the try-block would be easier to read here.

Reply:
Yes, good call, agreed!